riscv/csrs.cc

```diff
@@ -1421,17 +1421,17 @@ virtualized_stimecmp_csr_t::virtualized_stimecmp_csr_t(processor_t* const proc,
 }
 
 void virtualized_stimecmp_csr_t::verify_permissions(insn_t insn, bool write) const {
-  virtualized_csr_t::verify_permissions(insn, write);
-
-  // check for read permission to time as enabled by xcounteren
-  state->time_proxy->verify_permissions(insn, false);
+  const bool mtime_ok = (state->prv < PRV_M) ? (state->mcounteren->read() & 0x2) &&
+                                               (state->menvcfg->read() & MENVCFG_STCE)
+                                             : true;
+  const bool htime_ok = state->v ? (state->hcounteren->read() & 0x2) &&
+                                   (state->henvcfg->read() & HENVCFG_STCE)
+                                 : true;
+
+  if (!mtime_ok)
+    throw trap_illegal_instruction(insn.bits());
+  if (!htime_ok)
+      throw trap_virtual_instruction(insn.bits());
```

Comment:
Much of this is reimplementing the `time_proxy` permissions (from `counter_proxy_csr_t::verify_permissions()`), plus a few extra things. Isn't there some way to do this without copy/paste/tweak?

Can you check the higher-priority exceptions first, then call `time_proxy->verify_permissions()`?

Reply:
Yeah. Maybe we can do it like this:
```
  if (!(state->menvcfg->read() & MENVCFG_STCE)) {
  virtualized_csr_t::verify_permissions(insn, write);
    // access to (v)stimecmp with MENVCFG.STCE = 0
    if (state->prv < PRV_M)
      throw trap_illegal_instruction(insn.bits());
  } 

state->time_proxy->verify_permissions(insn, false);

 if (state->v && !(state->henvcfg->read() & HENVCFG_STCE)) {
    // access to vstimecmp with MENVCFG.STCE = 1 and HENVCFG.STCE = 0 when V = 1
    throw trap_virtual_instruction(insn.bits());
  }
```